src/main/java/org/jabref/model/database/BibDatabase.java

```diff
@@ -193,21 +194,32 @@ public synchronized boolean insertEntry(BibEntry entry) throws KeyCollisionExcep
      * @return false if the insert was done without a duplicate warning
      */
     public synchronized boolean insertEntry(BibEntry entry, EntryEventSource eventSource) throws KeyCollisionException {
-        Objects.requireNonNull(entry);
+        insertEntries(Collections.singletonList(entry), eventSource);
+        return duplicationChecker.isDuplicateCiteKeyExisting(entry);
+    }
 
-        String id = entry.getId();
-        if (containsEntryWithId(id)) {
-            throw new KeyCollisionException("ID is already in use, please choose another");
-        }
+    public synchronized void insertEntries(List<BibEntry> entries) throws KeyCollisionException {
+        insertEntries(entries, EntryEventSource.LOCAL);
+    }
 
-        internalIDs.add(id);
-        entries.add(entry);
-        entry.registerListener(this);
+    private synchronized void insertEntries(List<BibEntry> newEntries, EntryEventSource eventSource) throws KeyCollisionException {
+        Objects.requireNonNull(newEntries);
 
-        eventBus.post(new EntryAddedEvent(entry, eventSource));
-        return duplicationChecker.isDuplicateCiteKeyExisting(entry);
+        for (BibEntry entry : newEntries) {
+            String id = entry.getId();
+            if (containsEntryWithId(id)) {
```

Comment:
I'm not sure that this is really code is really the problem - `containsEntryWithId` is just a lookup in a relatively small (< 10.000) hashset. I have the feeling that our logic and model code is quite optimized or at least performs reasonable well but there are some problems with GUI code (e.g. adding 10 entries at the same time leads to 10 separate redraws of the maintable).